### PR TITLE
+ add method eventWrapper() to eliminate all self=this and curry returned function for event listeners

### PR DESCRIPTION
**Pull request by n0099**
* now listening more generic pointer events which including both touch and mouse events @ init
* rename all event listeners from `eventnameEvent` to `onEventName`
@ `plot/br-heatmap.ts`.BrHeatmap

### CODE REVIEW

**Review by ControlNet — COMMENTED** [review]
Line 49: Please remove unnecessary `console.log`.

Please consider to edit `/src/plot/line-chart-ts` as well.

**Review by ControlNet — CHANGES_REQUESTED** [review]
Type checking error. Please revise the code.

**Comment by ControlNet** [comment]
Need revise the event listeners in `line-chart.ts`.

File: src/plot/line-chart.ts [hunk 1]
```diff
@@ -184,7 +184,7 @@ export class BrLineChart extends LineChart {
         return this;
     }
 
-    brHeatmap(): BrHeatmap {
+    get brHeatmap(): BrHeatmap {
```
Comment (n0099) [inline src/plot/line-chart.ts:187]: all getters are useless since u are not defining a corresponding setter method

**Review by ControlNet — COMMENTED** [review]
image

It looks doesn't work on my end. Do you have any idea?

**Comment by n0099** [comment]
> image
> 
> It looks doesn't work on my end. Do you have any idea?

but `webpack build` will success without any ts error/warn

File: src/plot/br-heatmap.ts [hunk 2]
```diff
@@ -42,60 +44,54 @@ export class BrHeatmap extends Plot {
...
-        const self = this;
-        return function(): void {
-            d3.select(this).style("stroke", "black");
-            self.tooltip.hide();
-        };
+    static eventWrapper<S extends SVGRectElement | SVGSVGElement, T extends (d: SquareInfo, node: S) => void | Promise<void>>(cb: T)
+    : (this: S, d: SquareInfo, i: number, n: S[]) => void {
+        return (d, i, n) =>
+            // https://stackoverflow.com/questions/27746304/how-to-check-if-an-object-is-a-promise/27760489#27760489
+            Promise.resolve((cb.bind(this) as T)(d, n[i])).then(() => {});
```
Comment (ControlNet) [inline src/plot/br-heatmap.ts:51]: The tooltip doesn't work after updating. Please revise it.

br-heatmap.ts:69 Uncaught (in promise) TypeError: Cannot read properties of undefined (reading 'update')
    at Function.<anonymous> (br-heatmap.ts:69:28)
    at step (color-map-def.ts:44:68)
    at Object.next (color-map-def.ts:44:68)
    at color-map-def.ts:44:68
    at new Promise (<anonymous>)
    at __webpack_modules__../src/plot/br-heatmap.ts.__awaiter (color-map-def.ts:44:68)
    at __webpack_modules__../src/plot/br-heatmap.ts.BrHeatmap.onPointerMove (br-heatmap.ts:68:60)
    at SVGRectElement.<anonymous> (br-heatmap.ts:51:49)
    at SVGRectElement.<anonymous> (on.js:27:1)

**Review by ControlNet — CHANGES_REQUESTED** [review]
Need revision.

**Review by ControlNet — APPROVED** [review]

**Pull request merged**